internal/controller/vrg_volrep.go

```diff
@@ -2887,3 +2890,302 @@ func PruneAnnotations(annotations map[string]string) map[string]string {
+				"action", "manual cleanup required")
+
+			return pvcProcessResult{skip: true}
+		}
+
+		log.Info("Continue with cleanup",
+			"pvc", pvc.GetName(),
+			"ownedByVM", vm.GetName(),
+			"ownerReferences", pvc.OwnerReferences)
+
+		// Already valid; no patching needed
+		return pvcProcessResult{}
+	}
+
+	// 2. Check if PVC is used by virt-launcher pod
+	ownerVMMetadata, err := rmnutil.IsUsedByVirtLauncherPod(v.ctx, v.reconciler.Client, pvc, log)
+	if err != nil {
+		// Requeue to retry
+		return pvcProcessResult{retry: true}
+	}
+
+	if ownerVMMetadata == nil {
+		// skip automated cleanup, manual cleanup required
+		return pvcProcessResult{skip: true}
+	}
+
+	// Ensure the reported VM is of the expected type and protected
+	vmName := ownerVMMetadata.GetName()
+	if _, ok := protectedVMSet[vmName]; !ok {
+		log.Info("Skipping cleanup, PVC not used by protected VM",
+			"pvc", pvc.Name,
+			"vm", vmName)
+
+		return pvcProcessResult{skip: true}
+	}
+
+	vm, ok := ownerVMMetadata.(*metav1.PartialObjectMetadata)
+	if !ok {
+		log.Info("VirtualMachine(VM) map function received non-VM resource")
+
+		return pvcProcessResult{}
+	}
+
+	return pvcProcessResult{vm: vm}
+}
+
+// patchPVCsOwnerRefs applies VM OwnerReference to PVCs and returns true if any patch failed.
+func (v *VRGInstance) patchPVCsOwnerRefs(pvcToVM map[string]*metav1.PartialObjectMetadata) bool {
+	var patchError bool
+
+	for pvcName, vm := range pvcToVM {
+		err := rmnutil.PatchPvcWithVMOwnerRef(v.ctx, v.reconciler.Client, vm, pvcName, vm.Namespace, v.log)
+		if err != nil {
+			v.log.Error(err, "Failed to patch PVC owner reference",
+				"pvcName", pvcName,
+				"vm", vm.GetName())
+
+			patchError = true
+		}
+	}
+
+	return patchError
+}
+
+func toSet(items []string) map[string]struct{} {
+	m := make(map[string]struct{}, len(items))
+	for _, it := range items {
+		m[it] = struct{}{}
+	}
+
+	return m
+}
+
+// skip VM cleanup if cleanup is already in progress or cleanup just completed.
+// Returns TRUE to skip further validation and requeue for reconcile
+// Rerurns FALSE to proceed further with VM GarbageCollection process
+// Returns empty VM array list and FALSE if VM GC is complete and no need to requeue for reconcile
+func (v *VRGInstance) skipVMCleanupVerificationCheck() ([]virtv1.VirtualMachine, bool) {
+	vmNamespaceList := v.instance.Spec.KubeObjectProtection.RecipeParameters[recipecore.ProtectedVMNamespace]
+	vmList := v.instance.Spec.KubeObjectProtection.RecipeParameters[recipecore.VMList]
+
+	var foundVMs []virtv1.VirtualMachine
+
+	if len(vmList) == 0 {
+		v.log.V(1).Info("no protected VMs specified. Skipping VM garbage collection processing")
+
+		return nil, true
+	}
+
+	if rmnutil.IsVMDeletionInProgress(v.ctx, v.reconciler.Client, vmList, vmNamespaceList, v.log) {
+		v.log.Info("VM deletion is in progress, skipping ownerreferences check")
+
+		return foundVMs, true
+	}
+
+	foundVMs, err := rmnutil.ListVMsByVMNamespace(v.ctx, v.reconciler.APIReader,
```

Comment:
Why use the `APIReader` in this case? This causes additional calls to the API server, rather than using the cache, and so can cause the controller to hit API request limits. We usually use the reader sparsely. (NOTE: Applies to other callers of `ListVMsByVMNamespace` as well).